cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/KafkaMetadataStateManagerMockTest.java

```diff
@@ -0,0 +1,640 @@
+                .addToAnnotations(Annotations.ANNO_STRIMZI_IO_KRAFT, "rollback")
+                .endMetadata()
+                .withNewStatus()
+                .withKafkaMetadataState(KRaftMigration.name())
+                .endStatus()
+                .build();
+
+        ResourceOperatorSupplier supplier = ResourceUtils.supplierWithMocks(false);
+
+        SecretOperator mockSecretOps = supplier.secretOperations;
+        Secret secret = new Secret();
+        when(mockSecretOps.getAsync(eq(NAMESPACE), eq(KafkaResources.clusterCaCertificateSecretName(CLUSTER_NAME)))).thenReturn(Future.succeededFuture(secret));
+        when(mockSecretOps.getAsync(eq(NAMESPACE), eq(KafkaResources.secretName(CLUSTER_NAME)))).thenReturn(Future.succeededFuture(secret));
+
+        MockKafkaReconciler kr = new MockKafkaReconciler(
+                new Reconciliation("test-trigger", Kafka.RESOURCE_KIND, NAMESPACE, CLUSTER_NAME),
+                vertx,
+                CONFIG,
+                supplier,
+                new PlatformFeaturesAvailability(false, KUBERNETES_VERSION),
+                kafka,
+                List.of(BROKERS, CONTROLLERS),
+                createKafkaCluster(KAFKA),
+                CLUSTER_CA,
+                CLIENTS_CA);
+
+        KafkaStatus status = new KafkaStatus();
+
+        Checkpoint async = context.checkpoint();
+        kr.reconcile(status, Clock.systemUTC()).onComplete(res -> context.verify(() -> {
+            assertTrue(status.getConditions().stream().anyMatch(condition -> "KafkaMetadataStateWarning".equals(condition.getReason())));
+            assertEquals(status.getConditions().get(2).getMessage(), "The strimzi.io/kraft annotation can't be set to rollback during a migration process." +
+                    "It can be used in post migration to start rollback process.");
+            assertEquals(status.getKafkaMetadataState(), KRaftMigration.name());
+            async.flag();
+        }));
+    }
+
+    /** Mock for KafkaMetadataStateManager class */
+    static class MockKafkaMetadataStateManager extends KafkaMetadataStateManager {
+
+        /**
+         * Constructor
+         *
+         * @param reconciliation             Reconciliation information
+         * @param kafkaCr                    instance of the Kafka CR
+         * @param useKRaftFeatureGateEnabled if the UseKRaft feature gate is enabled on the operator
+         */
+        public MockKafkaMetadataStateManager(Reconciliation reconciliation, Kafka kafkaCr, boolean useKRaftFeatureGateEnabled) {
+            super(reconciliation, kafkaCr, useKRaftFeatureGateEnabled);
+        }
+
+        @Override
+        KafkaAgentClient initKafkaAgentClient(Reconciliation reconciliation, Secret clusterCaCertSecret, Secret coKeySecret) {
+            return mock(KafkaAgentClient.class, invocation -> {
+                if ("getKRaftMigrationState".equals(invocation.getMethod().getName())) {
+                    return new KRaftMigrationState(1);
```

Comment:
with this mock, we got that the migration is done as soon as the Kafka agent client is called. What if we want to test that the migration takes long, so for example we need 2-3 reconciliation to get it's done. I think we should cover this use case as well.